### PR TITLE
Delete related VSRs after restore completes.

### PR DESCRIPTION
**Pull request by mrnold**
[OADP-1872](https://issues.redhat.com//browse/OADP-1872): delete VSRs after a restore completes, and clean up VSRs from completed restores and restores that no longer exist.

### CODE REVIEW

**Comment by mrnold** [comment]
Okay, I think this satisfies all the requirements, and it's working okay. It is very slow though, so I will keep working to try to have it not hammer the API so hard.

File: pkg/datamover/datamover.go [hunk 1]
```diff
@@ -122,3 +123,174 @@ func CheckIfVolumeSnapshotBackupsAreComplete(ctx context.Context, volumesnapshot
...
+		return err
+	}
+
+	err = client.List(context.TODO(), &restoreList, &kbclient.ListOptions{})
+	if err != nil {
+		log.Errorf(err.Error())
+		return err
+	}
+
+	for _, restore := range restoreList.Items {
```
Comment (shubham-pampattiwar) [inline pkg/datamover/datamover.go]: We are calling the `CleanupRestoreVSRs` function for the restores with completed phase, IMO we should only cleanup VSRs for this current restore and not touch any other VSRs.

Comment (sseago) [reply pkg/datamover/datamover.go]: @shubham-pampattiwar Hmm. I guess that depends on whether we want to clean up old non-cleaned restores or not. If we're only ever concerned with removing VSRs for the current restore, then this should be simpler.

Comment (weshayutin) [reply pkg/datamover/datamover.go]: clean up only the vsr's associated w/ the restore in process.  Matching the vsb and vsr behavior would be the right thing to do.  Customers can clean any other $vsr's themselves.

Comment (mrnold) [reply pkg/datamover/datamover.go]: I put this back to only removing VSRs for the current restore.

File: pkg/controller/restore_controller.go [hunk 2]
```diff
@@ -575,6 +576,9 @@ func (r *restoreReconciler) runValidatedRestore(restore *api.Restore, info backu
 			r.logger.Debug("Restore completed")
 			restore.Status.Phase = api.RestorePhaseCompleted
 			r.metrics.RegisterRestoreSuccess(restore.Spec.ScheduleName)
+			if err := datamover.DeleteVSRsIfComplete(restore.Name, r.logger); err != nil {
+				r.logger.WithError(err).Error("Error removing VSRs after completed restore")
+			}
```
Comment (sseago) [inline pkg/controller/restore_controller.go:584]: We need the same above after line 568/569, otherwise a single restore error will prevent VSR cleanup.

Comment (sseago) [reply pkg/controller/restore_controller.go:584]: Also, what if there are inProgress operations, does the cleanup just never happen, or is there another place which will call the same code?

File: pkg/datamover/datamover.go [hunk 3]
```diff
@@ -122,3 +122,89 @@ func CheckIfVolumeSnapshotBackupsAreComplete(ctx context.Context, volumesnapshot
...
+			err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+				currentVSR := snapmoverv1alpha1.VolumeSnapshotRestore{}
+				err := volumeSnapMoverClient.Get(ctx, kbclient.ObjectKey{Namespace: volumesnapshotrestore.Namespace, Name: volumesnapshotrestore.Name}, &currentVSR)
+				if err != nil {
+					return false, errors.Wrapf(err, fmt.Sprintf("failed to get volumesnapshotrestore %s/%s", volumesnapshotrestore.Namespace, volumesnapshotrestore.Name))
+				}
+				// check for a failed VSR
+				for _, cond := range currentVSR.Status.Conditions {
+					if cond.Status == metav1.ConditionFalse && cond.Reason == ReconciledReasonError && cond.Type == ConditionReconciled && (currentVSR.Status.Phase == SnapMoverBackupPhaseFailed || currentVSR.Status.Phase == SnapMoverBackupPhasePartiallyFailed) {
+						return false, errors.Errorf("volumesnapshotrestore %s has failed status", currentVSR.Name)
```
Comment (sseago) [inline pkg/datamover/datamover.go:191]: Why do we return error if VSR failed? This causes the check command to fail out, which means we only clean up VSRs if they're all successful. Is this the desired behavior, or do we always want to clean up post-restore?

Comment (eemcmullan) [reply pkg/datamover/datamover.go:191]: @sseago I think we decided to only clean up completed CRs for debugging purposes

Comment (sseago) [reply pkg/datamover/datamover.go:191]: Disregard this comment, as we want the current behavior here.

File: pkg/controller/restore_controller.go (hunk 2 above)
Comment (mrnold) [reply pkg/controller/restore_controller.go:584]: I think I have addressed these with [cc1c0](https://github.com/openshift/velero/pull/269/commits/cc1c0c2c7ed1a7a23e8594d64a24b67ad56103fd) and [7e77a](https://github.com/openshift/velero/pull/269/commits/7e77a30392be0921717a8c1da185f3ca448c282a) now, let me know whether or not that's what you had in mind.

**Comment by shubham-pampattiwar** [comment]
@mrnold Unit tests are failing please make sure you sun `make update` and make test` locally and see if there are any changes needed

File: pkg/controller/restore_operations_controller.go [hunk 4]
```diff
@@ -186,14 +187,20 @@ func (r *restoreOperationsReconciler) Reconcile(ctx context.Context, req ctrl.Re
...
 			r.metrics.RegisterRestoreSuccess(restore.Spec.ScheduleName)
+			phase = " completed "
 		} else {
 			log.Infof("Marking restore %s FinalizingPartiallyFailed", restore.Name)
 			restore.Status.Phase = velerov1api.RestorePhasePartiallyFailed
 			r.metrics.RegisterRestorePartialFailure(restore.Spec.ScheduleName)
+			phase = " partially failed "
+		}
+		if err = datamover.DeleteVSRsIfComplete(restore.Name, log); err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "error cleaning up after%srestore", phase)
```
Comment (sseago) [inline pkg/controller/restore_operations_controller.go]: I don't think we really need the extra phase text in the message. That makes the flow slightly more confusing, and it doesn't really add much.

**Comment by weshayutin** [comment]
The unit test https://github.com/openshift/velero/blob/konveyor-dev/pkg/controller/restore_operations_controller_test.go#L70

https://github.com/openshift/velero/blob/konveyor-dev/pkg/controller/restore_operations_controller_test.go#L308

is failing via cli invocation, goland invocation but passing vscode.
this is silly.. 
waiving test for 1.2 branch

**Comment by weshayutin** [comment]
@sseago @kaovilai @mrnold @shubham-pampattiwar on a call waiving.

**Pull request merged**